### DISCUSSION TITLE
Cache miss when I pod update some sdk

### DISCUSSION THREAD

**Discussion by wolfcolony** [ccache/ccache#501] [Support]
### Question ###

My project not use Ccache  build time probably 650s firstly，use Ccache build time probably 1300s firstly,  when i not update some SDK build on second time，it hit cache probably 70%+，then build time will reduce  to 350s，but if i pod update some SDK，Ccache will hit cache missed，and it cause build time too long, may be 800s or even worse，i dont know why some SDK code not modify only update other SDK，cause this situation

**Comment by jrosdahl** [comment]
Hi there.

It's next to impossible to give you any advice from the information you supplied. For instance: What is "pod update"? What do you mean with "some SDK" and "other SDK"? Which version of ccache did you use? Which OS do you use? What have you tried? What did you expect?

I suggest you to read the [Cache debugging](https://ccache.dev/manual/3.7.6.html#_cache_debugging) section of the manual. I also recommend [Effective Ways to Get Help from Maintainers](https://www.snoyman.com/blog/2017/10/effective-ways-help-from-maintainers) for tips on how to ask questions in an effective way.

**Comment by wolfcolony** [comment]
Sorry for my description which is not clearness. My project is Objective-C Cocoa projects and the platform I'm using is **macOS** + **Xcode** + **Clang**. I used the latest version **Ccache** instead of xcode's default compilation instruction **clang** for build more faster.

 Xcode projects typically use [**CocoaPods** ](https://cocoapods.org/) which is a third-party dependency manager for Objective-C Cocoa projects. I found that whenever I used CocoaPods pod update some third-party code would cause Cache miss.

For instance: ClassA has an third-party SDKB dependency.  When I updated the SDKB even though the ClassA code did not change. It will cause build ClassA cache miss and ccache-clang build too long. Is there any way to ensure that ClassA 's cache does not miss when the SDKB is updated? Thank You.

**Comment by wolfcolony** [comment]
I wrote a demo like this but it used **clang** build：

```
#import "ClassA.h"
#import "SDKB.h"

@implementation ClassA

@end
```
I build it and look the ClassA.o binary data like this：

```
cffa edfe 0c00 0001 0000 0000 0100 0000
0800 0000 5007 0000 0020 0000 0000 0000
1900 0000 3806 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
fd06 0000 0000 0000 7007 0000 0000 0000
...
```
and then I modify SDKB code rebuild look the ClassA.o binary data like this：

```
cffa edfe 0c00 0001 0000 0000 0100 0000
0800 0000 5007 0000 0020 0000 0000 0000
1900 0000 3806 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
fd06 0000 0000 0000 7007 0000 0000 0000
...
```
It seems that no matter how much I change the SDKB but ClassA.o doesn't change. In my opinion if the ClassA.o does not change, the cache generated by **Ccache** should not be invalidated, should it? Looking forward to your reply, Thank You.

**Comment by wolfcolony** [comment]
And my Ccache configuration is like this：
```
export CCACHE_MAXSIZE=10G
export CCACHE_CPP2=true
export CCACHE_HARDLINK=true
export CCACHE_COMPILERCHECK=content
export CCACHE_SLOPPINESS=file_macro,time_macros,pch_defines
# export CCACHE_LOGFILE=~/Desktop/CCache.log

```

**Comment by jrosdahl** [comment]
(Minor note: It's "ccache" or "Ccache", not "CCache". :wink:)

> It seems that no matter how much I change the SDKB but ClassA.o doesn't change. In my opinion if the ClassA.o does not change, the cache generated by Ccache should not be invalidated, should it?

That's impossible for me to say given the current information. But there is a method that _you_ can use to find out why ccache did not reuse the previously cached results: read the [Cache debugging](https://ccache.dev/manual/3.7.7.html#_cache_debugging) section of the manual and perform the steps under _To debug why you don’t get an expected cache hit for an object file, you can do something like this_ and you will get useful data. If you find it hard to interpret the result you can supply the log and "ccache-input-text" diff here.

**Comment by wolfcolony** [comment]
> (Minor note: It's "ccache" or "Ccache", not "CCache". 😉)
> 
> > It seems that no matter how much I change the SDKB but ClassA.o doesn't change. In my opinion if the ClassA.o does not change, the cache generated by Ccache should not be invalidated, should it?
> 
> That's impossible for me to say given the current information. But there is a method that _you_ can use to find out why ccache did not reuse the previously cached results: read the [Cache debugging](https://ccache.dev/manual/3.7.7.html#_cache_debugging) section of the manual and perform the steps under _To debug why you don’t get an expected cache hit for an object file, you can do something like this_ and you will get useful data. If you find it hard to interpret the result you can supply the log and "ccache-input-text" diff here.

Thank you for your advice. I tried it out in your way and compare the data of two files, I found It appears that the ~~PCH~~ Header file  affected the cache hit, I wrote an [example](https://github.com/wolfcolony/ExampleLog.git) where I detailed the cache misses and the output file [before](https://github.com/wolfcolony/ExampleLog/tree/master/OldLog) and [after](https://github.com/wolfcolony/ExampleLog/tree/master/NewLog) the modification is added.

I find that if modify SDB.h content , ClassA.o.ccache-input-text's PREPROCESSOR MODE content will changed and cause cache miss. but use **clang** ClassA.o hasn't changed at all. Is this the bug of ccache or my usage way is wrong.

I don't know if there are any other ccache configurations that can avoid the effects of PCH, but I tried many of them without any effect. Even if I didn't change the source file, just change the header file of include without causing cache invalidation. I wonder if there is a way to avoid or even turn off PREPROCESSOR MODE and just cache the source code. Looking forward to your reply, Thank You.

**Comment by jrosdahl** [comment]
> I find that if modify SDB.h content

In what way did you modify it? There are many ways to modify source code that result in different preprocessed output but the same object file.

> ClassA.o.ccache-input-text's PREPROCESSOR MODE content will changed and cause cache miss. but use clang ClassA.o hasn't changed at all. Is this the bug of ccache or my usage way is wrong.

That's hard to tell since there is no log of what the preprocessed code looked like when the result was placed in the cache the first time. Try cleaning the cache and then catching the log from a cache miss and a log of another cache miss.
